test/Mono.Linker.Tests.Cases/Reflection/TypeHierarchyReflectionWarnings.cs

```diff
@@ -0,0 +1,444 @@
+			// Reference to the derived type should apply base annotations
+			var t = typeof (DerivedFromAnnotatedBase);
+			RequirePublicMethods (annotatedDerivedFromBase.GetType ());
+			RequirePublicNestedTypes (annotatedPublicNestedTypes.GetType ());
+			RequirePublicFields (derivedFromAnnotatedDerivedFromBase.GetType ());
+			RequirePublicMethods (annotatedPublicMethods.GetType ());
+			RequirePublicFields (annotatedPublicFields.GetType ());
+			RequirePublicProperties (annotatedPublicProperties.GetType ());
+			RequirePublicEvents (annotatedPublicEvents.GetType ());
+			RequirePublicNestedTypes (annotatedPublicNestedTypes.GetType ());
+			RequireInterfaces (annotatedInterfaces.GetType ());
+			RequireAll (annotatedAll.GetType ());
+			RequirePublicMethods (annotatedRUCPublicMethods.GetType ());
+
+			// Instantiate this type just so its property getters are considered reachable
+			var b = new DerivedFromAnnotatedDerivedFromBase ();
+		}
+
+		[Kept]
+		static AnnotatedAll annotatedAll;
+		[Kept]
+		static AnnotatedPublicMethods annotatedPublicMethods;
+		[Kept]
+		static AnnotatedPublicFields annotatedPublicFields;
+		[Kept]
+		static AnnotatedPublicProperties annotatedPublicProperties;
+		[Kept]
+		static AnnotatedPublicEvents annotatedPublicEvents;
+		[Kept]
+		static AnnotatedInterfaces annotatedInterfaces;
+		[Kept]
+		static AnnotatedBase annotatedBase;
+		[Kept]
+		static AnnotatedDerivedFromBase annotatedDerivedFromBase;
+		[Kept]
+		static AnnotatedPublicNestedTypes annotatedPublicNestedTypes;
+		[Kept]
+		static DerivedFromAnnotatedDerivedFromBase derivedFromAnnotatedDerivedFromBase;
+		[Kept]
+		static AnnotatedRUCPublicMethods annotatedRUCPublicMethods;
+
+		[Kept]
+		[KeptMember (".ctor()")]
+		[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+		[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.All)]
+		class AnnotatedAll
+		{
+			[Kept]
+			[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+			[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+			[ExpectedWarning ("IL2114", nameof (AnnotatedAll), nameof (DAMTField))]
+			public Type DAMTField;
+
+			[Kept]
+			[KeptAttributeAttribute (typeof (RequiresUnreferencedCodeAttribute))]
+			[ExpectedWarning ("IL2112", "--RUC on AnnotatedAll.RUCMethod--")]
+			[RequiresUnreferencedCode ("--RUC on AnnotatedAll.RUCMethod--")]
+			public void RUCMethod () { }
+
+			[Kept]
+			[ExpectedWarning ("IL2114", nameof (AnnotatedAll), nameof (DAMTMethod))]
+			public void DAMTMethod (
+				[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+				[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+				Type t
+			)
+			{ }
+		}
+
+		[Kept]
+		[KeptAttributeAttribute (typeof (DynamicallyAccessedMembersAttribute))]
+		[DynamicallyAccessedMembers (DynamicallyAccessedMemberTypes.PublicMethods)]
+		class AnnotatedPublicMethods
+		{
+			[Kept]
```

Comment:
I would add a DAM field and check that it doesn't produce a warning in this case (if you already have test cases for All and PublicMethods and so on).

Reply:
Done!